File: benchmarks/semantic_search/fixtures/go-workflow-discovery-v1/internal/selection/retry.go
package selection

import (
	"context"

	"example.com/synthetic-workflow-discovery/internal/discovery"
)

func SelfCorrectSelection(ctx context.Context, validator *Validator, candidate string) error {
	state, ok := discovery.WorkflowStateFromContext(ctx)
	if ok {
		validator.SetWorkflowState(state)
	}
	return validator.Validate(candidate)
}

func SelectWithRetry(ctx context.Context, validator *Validator, candidate string) error {
	if err := SelfCorrectSelection(ctx, validator, candidate); err != nil {
		return retrySelection(ctx, validator, candidate)
	}
	return nil
}

func retrySelection(ctx context.Context, validator *Validator, candidate string) error {
	return SelfCorrectSelection(ctx, validator, candidate)
}

File: benchmarks/semantic_search/fixtures/python-workflow-discovery-v1/selection/retry.py
from __future__ import annotations

from selection.validator import Validator


def select_with_retry(validator: Validator, workflow_id: str, corrected_id: str) -> str:
    try:
        validator.validate(workflow_id)
        return workflow_id
    except Exception:
        return self_correct_selection(validator, corrected_id)


def self_correct_selection(validator: Validator, workflow_id: str) -> str:
    validator.validate(workflow_id)
    return workflow_id
